// sanity-plugin-bynder-input/dist/sanity-plugin-bynder-input.esm.js
import React, { useRef, useState, useEffect } from 'react';
import PatchEvent, { unset, set } from '@sanity/form-builder/PatchEvent';
import pluginConfig from 'config:bynder-input';
import ButtonGrid from 'part:@sanity/components/buttons/button-grid';
import Button from 'part:@sanity/components/buttons/default';
import Fieldset from 'part:@sanity/components/fieldsets/default';
import videojs from 'video.js';
import { DiffFromTo } from '@sanity/field/diff';

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

var url = 'https://ucv.bynder.com/5.0.5/modules/compactview/bynder-compactview-3-latest.js';
function loadBynder(callback) {
  var existingScript = document.getElementById('bynder');

  if (!existingScript) {
    var script = document.createElement('script');
    script.src = url;
    script.id = 'bynder';
    document.body.appendChild(script);

    script.onload = function () {
      if (callback) {
        return callback();
      }

      return true;
    };
  }

  if (existingScript && callback) {
    return callback();
  }

  return true;
}

var VideoPlayer = function VideoPlayer(props) {
  var videoNode = useRef(null);

  var _useState = useState(),
      player = _useState[0],
      setPlayer = _useState[1];

  useEffect(function () {
    if (videoNode.current) {
      setPlayer(videojs(videoNode.current, props));
    }

    return function () {
      return player == null ? void 0 : player.dispose();
    };
  }, [videoNode]);
  return React.createElement("div", null, React.createElement("link", {
    href: "https://vjs.zencdn.net/7.8.4/video-js.css",
    rel: "stylesheet"
  }), React.createElement("div", {
    "data-vjs-player": true,
    style: {
      marginBottom: '16px'
    }
  }, React.createElement("video", {
    className: "video-js vjs-16-9 vjs-big-play-centered",
    ref: videoNode
  })));
};

var BynderInput = function BynderInput(props) {
  var removeValue = function removeValue() {
    var onChange = props.onChange;
    onChange(PatchEvent.from([unset()]));
  };

  var getPreviewUrl = function getPreviewUrl(asset) {
    switch (asset.type) {
      case 'VIDEO':
        return asset.previewUrls[0];

      default:
        return asset.files.webImage.url;
    }
  };

  var getVideoUrl = function getVideoUrl(asset) {
    if (asset.type === 'VIDEO') {
      var _asset$files$original, _asset$files, _asset$files$original2;

      // if original asset is available (public videos only) use that if not fall back to the preview url
      return (_asset$files$original = (_asset$files = asset.files) == null ? void 0 : (_asset$files$original2 = _asset$files.original) == null ? void 0 : _asset$files$original2.url) != null ? _asset$files$original : asset.previewUrls[0];
    }

    return null;
  };

  var getAspectRatio = function getAspectRatio(dimensions) {
    return dimensions.height / dimensions.width;
  };

  var getVideoAspectRatio = function getVideoAspectRatio(previewImageUrl) {
    return new Promise(function (resolve, reject) {
      var img = new Image();

      img.onload = function () {
        resolve(getAspectRatio({
          width: img.width,
          height: img.height
        }));
      };

      img.onerror = function (err) {
        return reject(err);
      };

      img.src = previewImageUrl;
    });
  };

  var openMediaSelector = function openMediaSelector() {
    var onChange = props.onChange,
        type = props.type,
        value = props.value;

    var onSuccess = function onSuccess(assets) {
      var _asset$files$transfor;

      var asset = assets[0];
      var webImage = asset.files.webImage;
      var aspectRatio = getAspectRatio({
        width: webImage.width,
        height: webImage.height
      });
      var mediaData = {
        _key: value == null ? void 0 : value._key,
        _type: type.name,
        id: asset.id,
        name: asset.name,
        databaseId: asset.databaseId,
        type: asset.type,
        previewUrl: getPreviewUrl(asset),
        previewImg: webImage.url,
        datUrl: (_asset$files$transfor = asset.files.transformBaseUrl) == null ? void 0 : _asset$files$transfor.url,
        videoUrl: getVideoUrl(asset),
        description: asset.description,
        aspectRatio: aspectRatio
      };

      if (asset.type === 'VIDEO') {
        getVideoAspectRatio(webImage.url).then(function (ratio) {
          onChange(PatchEvent.from([set(_extends({}, mediaData, {
            aspectRatio: ratio
          }))]));
        })["catch"](function (err) {
          // video aspect ratio couldn't be set, but should still set the rest of the data
          console.log('Error setting video aspect ratio:', err);
          onChange(PatchEvent.from([set(mediaData)]));
        });
      } else {
        onChange(PatchEvent.from([set(mediaData)]));
      }
    };

    var options = {
      language: pluginConfig.language || 'en_US',
      portal: {
        url: pluginConfig.portalDomain
      },
      theme: {
        colorPrimary: '#156dff',
        colorButtonPrimary: '#156dff'
      },
      mode: 'SingleSelectFile',
      onSuccess: onSuccess
    };
    var assetTypes = type.options.assetTypes;

    if (assetTypes) {
      options.assetTypes = assetTypes;
    }

    loadBynder(function () {
      window.BynderCompactView.open(options);
    });
  };

  var value = props.value,
      type = props.type,
      markers = props.markers,
      level = props.level,
      readOnly = props.readOnly;
  var preview;

  if (value) {
    if (value.type === 'VIDEO') {
      preview = React.createElement(VideoPlayer, {
        controls: true,
        poster: value.previewImg,
        sources: [{
          src: value.previewUrl
        }]
      });
    }

    if (value.type === 'IMAGE') {
      preview = React.createElement("img", {
        alt: "preview",
        src: value.previewUrl,
        style: {
          maxWidth: '100%',
          height: 'auto'
        }
      }); // TODO: Add preview for document / audio types and empty state
    }
  }

  return React.createElement(Fieldset, {
    markers: markers,
    //presence={presence.filter(item => item.path[0] === '$' || isInside.includes(item.identity))}
    legend: type.title,
    description: type.description,
    level: level
  }, React.createElement("div", {
    style: {
      textAlign: 'center'
    }
  }, preview), React.createElement(ButtonGrid, {
    align: "start"
  }, React.createElement(Button, {
    disabled: readOnly,
    inverted: true,
    title: "Select an asset",
    kind: "default",
    onClick: openMediaSelector
  }, "Browse"), React.createElement(Button, {
    disabled: readOnly || !value,
    color: "danger",
    inverted: true,
    title: "Remove asset",
    onClick: removeValue
  }, "Remove")));
};

var css_248z = "@import \"part:@sanity/base/theme/variables-style\";.root{display:flex;flex-direction:column;height:100%}.header{flex:1;min-height:0;padding:var(--small-padding) var(--small-padding) 0}.imageWrapper{background-color:color(var(--gray-lightest) alpha(50%));background-image:linear-gradient(45deg,var(--checkerboard-color) 25%,transparent 25%),linear-gradient(-45deg,var(--checkerboard-color) 25%,transparent 25%),linear-gradient(45deg,transparent 75%,var(--checkerboard-color) 75%),linear-gradient(-45deg,transparent 75%,var(--checkerboard-color) 75%);background-position:0 0,0 8px,8px -8px,-8px 0;background-size:16px 16px;display:flex;flex-direction:column;height:100%;max-height:190px;position:relative;&:after{bottom:0;box-shadow:inset 0 0 0 1px var(--hairline-color);content:\"\";display:block;left:0;pointer-events:none;position:absolute;right:0;top:0}}.image{display:block;flex:1;height:100%;min-height:0;-o-object-fit:contain;object-fit:contain;width:100%;@nest &[data-action=\"removed\"]{opacity:.5}}.error{composes:image;margin:50px 0;text-align:center}.hotspotCrop{display:block;height:100%;left:0;position:absolute;top:0;width:100%}.imageWrapperChanged{composes:imageWrapper;opacity:.45}.strikethrough{text-decoration:line-through}.noImage{composes:small from \"part:@sanity/base/theme/typography/text-blocks-style\";align-items:center;background:var(--gray-lightest);border-radius:var(--border-radius-small);color:var(--gray-dark);display:flex;height:100%;justify-content:center;width:100%}";

function Component(_ref) {
  var value = _ref.value;

  if (value && value.previewUrl) {
    return React.createElement("div", null, React.createElement("img", {
      alt: "preview",
      src: value.previewUrl,
      style: {
        width: '100%'
      }
    }));
  }

  return React.createElement("div", {
    className: css_248z.noImage
  }, React.createElement("div", null, "(no image)"));
}

var BynderDiff = function BynderDiff(_ref2) {
  var diff = _ref2.diff,
      schemaType = _ref2.schemaType;
  return React.createElement(DiffFromTo, {
    diff: diff,
    schemaType: schemaType,
    previewComponent: Component
  });
};

var schema = {
  name: 'bynder.asset',
  type: 'object',
  title: 'Bynder Asset',
  fields: [{
    type: 'string',
    name: 'id'
  }, {
    type: 'string',
    name: 'name'
  }, {
    type: 'string',
    name: 'databaseId'
  }, {
    type: 'string',
    name: 'type'
  }, {
    type: 'string',
    name: 'previewUrl'
  }, {
    type: 'string',
    name: 'previewImg'
  }, {
    type: 'string',
    name: 'datUrl'
  }, {
    type: 'string',
    name: 'description'
  }, {
    type: 'number',
    name: 'aspectRatio'
  }, {
    type: 'string',
    name: 'videoUrl'
  }],
  inputComponent: BynderInput,
  diffComponent: BynderDiff
};

export default schema;
