// sanity-plugin-bynder-input/dist/sanity-plugin-bynder-input.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PatchEvent = require('@sanity/form-builder/PatchEvent');
var PatchEvent__default = _interopDefault(PatchEvent);
var pluginConfig = _interopDefault(require('config:bynder-input'));
var ButtonGrid = _interopDefault(require('part:@sanity/components/buttons/button-grid'));
var Button = _interopDefault(require('part:@sanity/components/buttons/default'));
var Fieldset = _interopDefault(require('part:@sanity/components/fieldsets/default'));
var videojs = _interopDefault(require('video.js'));
var diff = require('@sanity/field/diff');

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

var url = 'https://ucv.bynder.com/5.0.5/modules/compactview/bynder-compactview-3-latest.js';
function loadBynder(callback) {
  var existingScript = document.getElementById('bynder');

  if (!existingScript) {
    var script = document.createElement('script');
    script.src = url;
    script.id = 'bynder';
    document.body.appendChild(script);

    script.onload = function () {
      if (callback) {
        return callback();
      }

      return true;
    };
  }

  if (existingScript && callback) {
    return callback();
  }

  return true;
}

var VideoPlayer = function VideoPlayer(props) {
  var videoNode = React.useRef(null);

  var _useState = React.useState(),
      player = _useState[0],
      setPlayer = _useState[1];

  React.useEffect(function () {
    if (videoNode.current) {
      setPlayer(videojs(videoNode.current, props));
    }

    return function () {
      return player == null ? void 0 : player.dispose();
    };
  }, [videoNode]);
  return React__default.createElement("div", null, React__default.createElement("link", {
    href: "https://vjs.zencdn.net/7.8.4/video-js.css",
    rel: "stylesheet"
  }), React__default.createElement("div", {
    "data-vjs-player": true,
    style: {
      marginBottom: '16px'
    }
  }, React__default.createElement("video", {
    className: "video-js vjs-16-9 vjs-big-play-centered",
    ref: videoNode
  })));
};

var BynderInput = function BynderInput(props) {
  var removeValue = function removeValue() {
    var onChange = props.onChange;
    onChange(PatchEvent__default.from([PatchEvent.unset()]));
  };

  var getPreviewUrl = function getPreviewUrl(asset) {
    switch (asset.type) {
      case 'VIDEO':
        return asset.previewUrls[0];

      default:
        return asset.files.webImage.url;
    }
  };

  var getVideoUrl = function getVideoUrl(asset) {
    if (asset.type === 'VIDEO') {
      var _asset$files$original, _asset$files, _asset$files$original2;

      // if original asset is available (public videos only) use that if not fall back to the preview url
      return (_asset$files$original = (_asset$files = asset.files) == null ? void 0 : (_asset$files$original2 = _asset$files.original) == null ? void 0 : _asset$files$original2.url) != null ? _asset$files$original : asset.previewUrls[0];
    }

    return null;
  };

  var getAspectRatio = function getAspectRatio(dimensions) {
    return dimensions.height / dimensions.width;
  };

  var getVideoAspectRatio = function getVideoAspectRatio(previewImageUrl) {
    return new Promise(function (resolve, reject) {
      var img = new Image();

      img.onload = function () {
        resolve(getAspectRatio({
          width: img.width,
          height: img.height
        }));
      };

      img.onerror = function (err) {
        return reject(err);
      };

      img.src = previewImageUrl;
    });
  };

  var openMediaSelector = function openMediaSelector() {
    var onChange = props.onChange,
        type = props.type,
        value = props.value;

    var onSuccess = function onSuccess(assets) {
      var _asset$files$transfor;

      var asset = assets[0];
      var webImage = asset.files.webImage;
      var aspectRatio = getAspectRatio({
        width: webImage.width,
        height: webImage.height
      });
      var mediaData = {
        _key: value == null ? void 0 : value._key,
        _type: type.name,
        id: asset.id,
        name: asset.name,
        databaseId: asset.databaseId,
        type: asset.type,
        previewUrl: getPreviewUrl(asset),
        previewImg: webImage.url,
        datUrl: (_asset$files$transfor = asset.files.transformBaseUrl) == null ? void 0 : _asset$files$transfor.url,
        videoUrl: getVideoUrl(asset),
        description: asset.description,
        aspectRatio: aspectRatio
      };

      if (asset.type === 'VIDEO') {
        getVideoAspectRatio(webImage.url).then(function (ratio) {
          onChange(PatchEvent__default.from([PatchEvent.set(_extends({}, mediaData, {
            aspectRatio: ratio
          }))]));
        })["catch"](function (err) {
          // video aspect ratio couldn't be set, but should still set the rest of the data
          console.log('Error setting video aspect ratio:', err);
          onChange(PatchEvent__default.from([PatchEvent.set(mediaData)]));
        });
      } else {
        onChange(PatchEvent__default.from([PatchEvent.set(mediaData)]));
      }
    };

    var options = {
      language: pluginConfig.language || 'en_US',
      portal: {
        url: pluginConfig.portalDomain
      },
      theme: {
        colorPrimary: '#156dff',
        colorButtonPrimary: '#156dff'
      },
      mode: 'SingleSelectFile',
      onSuccess: onSuccess
    };
    var assetTypes = type.options.assetTypes;

    if (assetTypes) {
      options.assetTypes = assetTypes;
    }

    loadBynder(function () {
      window.BynderCompactView.open(options);
    });
  };

  var value = props.value,
      type = props.type,
      markers = props.markers,
      level = props.level,
      readOnly = props.readOnly;
  var preview;

  if (value) {
    if (value.type === 'VIDEO') {
      preview = React__default.createElement(VideoPlayer, {
        controls: true,
        poster: value.previewImg,
        sources: [{
          src: value.previewUrl
        }]
      });
    }

    if (value.type === 'IMAGE') {
      preview = React__default.createElement("img", {
        alt: "preview",
        src: value.previewUrl,
        style: {
          maxWidth: '100%',
          height: 'auto'
        }
      }); // TODO: Add preview for document / audio types and empty state
    }
  }

  return React__default.createElement(Fieldset, {
    markers: markers,
    //presence={presence.filter(item => item.path[0] === '$' || isInside.includes(item.identity))}
    legend: type.title,
    description: type.description,
    level: level
  }, React__default.createElement("div", {
    style: {
      textAlign: 'center'
    }
  }, preview), React__default.createElement(ButtonGrid, {
    align: "start"
  }, React__default.createElement(Button, {
    disabled: readOnly,
    inverted: true,
    title: "Select an asset",
    kind: "default",
    onClick: openMediaSelector
  }, "Browse"), React__default.createElement(Button, {
    disabled: readOnly || !value,
    color: "danger",
    inverted: true,
    title: "Remove asset",
    onClick: removeValue
  }, "Remove")));
};

function Component(_ref) {
  var value = _ref.value;

  if (value && value.previewUrl) {
    return React__default.createElement("div", null, React__default.createElement("img", {
      alt: "preview",
      src: value.previewUrl,
      style: {
        width: '100%'
      }
    }));
  }

  return React__default.createElement("div", {
    className: undefined.noImage
  }, React__default.createElement("div", null, "(no image)"));
}

var BynderDiff = function BynderDiff(_ref2) {
  var diff$1 = _ref2.diff,
      schemaType = _ref2.schemaType;
  return React__default.createElement(diff.DiffFromTo, {
    diff: diff$1,
    schemaType: schemaType,
    previewComponent: Component
  });
};

var schema = {
  name: 'bynder.asset',
  type: 'object',
  title: 'Bynder Asset',
  fields: [{
    type: 'string',
    name: 'id'
  }, {
    type: 'string',
    name: 'name'
  }, {
    type: 'string',
    name: 'databaseId'
  }, {
    type: 'string',
    name: 'type'
  }, {
    type: 'string',
    name: 'previewUrl'
  }, {
    type: 'string',
    name: 'previewImg'
  }, {
    type: 'string',
    name: 'datUrl'
  }, {
    type: 'string',
    name: 'description'
  }, {
    type: 'number',
    name: 'aspectRatio'
  }, {
    type: 'string',
    name: 'videoUrl'
  }],
  inputComponent: BynderInput,
  diffComponent: BynderDiff
};

exports.default = schema;
